File: 43. Multiply Strings/Optimal.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        if(num1=="0" || num2=="0"){
            return "0";
        }
        int m = num1.size(), n = num2.size();
        vector<int> p(m+n,0);
        for(int i=m-1; i>=0; i--){
            for(int j=n-1; j>=0; j--){
                int multi = (num1[i]-'0') * (num2[j]-'0');
                int sum = multi + p[i+j+1];
                p[i+j+1] = sum%10;
                p[i+j] += sum/10;
            }
        }
        string ans = "";
        bool lead_z = true;
        for(int numb : p){
            if(numb != 0){
                lead_z = false;
            }
            if(!lead_z){
                ans += (numb + '0');
            }
        }
        return ans;
    }
};

File: 43. Multiply Strings/Optimal.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if num1=="0" or num2=="0":
            return "0"
        m,n = len(num1),len(num2)
        res = [0] * (m+n)
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                mul = int(num1[i]) * int(num2[j])
                suma = mul + res[i+j+1]
                res[i+j+1] = suma%10
                res[i+j] += suma//10
        ans = ""
        l = True
        for i in res:
            if i != 0:
                l = False
            if not l:
                ans += str(i)
        return ans
